notes/connection-lifecycle.md

```diff
@@ -0,0 +1,77 @@
+# Connection lifecycle
+
+## Current implementation
+
+`HTTPConnection` should be instantiated with `host` and `port` of the
+**first origin being connected to** to reach the target origin. This either means
+the target origin itself or the proxy origin if one is desired.
+
+If the connection is an `HTTPSConnection`, then `HTTPSConnection.set_cert()` should
+be called once with information about how to verify certificates. (What is stopping us
+from putting this information in `HTTPSConnection.__init__()`?)
+
+If using CONNECT tunneling with the proxy, call `HTTPConnection.set_tunnel()`
+with the tunneled host, port, and headers. This will mark the `HTTPConnection`
+as not-reusable.
+
+(If CONNECT tunneling and TLS-in-TLS is required, the HTTPSConnectionPool also sets `HTTPConnection.tls_in_tls_required` to true, is this required? Can the connection detect this from `proxy` and `proxy_config` reliably?)
```

Comment:
Great point, the connection also has access to `proxy` as `self.proxy`, so we can add the exact same condition in the connection itself.